src/Microsoft.Data/BaseColumn.cs

```diff
@@ -64,24 +67,26 @@ protected set
         /// <returns></returns>
         public virtual BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false) => throw new NotImplementedException();
 
+        public virtual BaseColumn CloneAndAppendNulls(long numberOfNulls, BaseColumn mapIndices = null, bool invertMapIndices = false) => throw new NotImplementedException();
+
         public virtual BaseColumn Sort(bool ascending = true) => throw new NotImplementedException();
 
         internal virtual BaseColumn GetAscendingSortIndices() => throw new NotImplementedException();
 
         internal delegate long GetBufferSortIndex(int bufferIndex, int sortIndex);
-        internal delegate T GetValueAtBuffer<T>(int bufferIndex, int valueIndex);
+        internal delegate ValueTuple<T, int> GetValueAndBufferSortIndexAtBuffer<T>(int bufferIndex, int valueIndex);
         internal delegate int GetBufferLengthAtIndex(int bufferIndex);
-        internal void PopulateColumnSortIndicesWithHeap<T>(SortedDictionary<T, List<Tuple<int, int>>> heapOfValueAndListOfTupleOfSortAndBufferIndex,
+        internal void PopulateColumnSortIndicesWithHeap<T>(SortedDictionary<T, List<ValueTuple<int, int>>> heapOfValueAndListOfTupleOfSortAndBufferIndex,
                                                             PrimitiveColumn<long> columnSortIndices,
                                                             GetBufferSortIndex getBufferSortIndex,
-                                                            GetValueAtBuffer<T> getValueAtBuffer,
+                                                            GetValueAndBufferSortIndexAtBuffer<T> getValueAndBufferSortIndexAtBuffer,
                                                             GetBufferLengthAtIndex getBufferLengthAtIndex)
         {
             while (heapOfValueAndListOfTupleOfSortAndBufferIndex.Count > 0)
             {
-                KeyValuePair<T, List<Tuple<int, int>>> minElement = heapOfValueAndListOfTupleOfSortAndBufferIndex.ElementAt(0);
-                List<Tuple<int, int>> tuplesOfSortAndBufferIndex = minElement.Value;
-                Tuple<int, int> sortAndBufferIndex;
+                KeyValuePair<T, List<ValueTuple<int, int>>> minElement = heapOfValueAndListOfTupleOfSortAndBufferIndex.ElementAt(0);
+                List<ValueTuple<int, int>> tuplesOfSortAndBufferIndex = minElement.Value;
+                (int sortIndex, int bufferIndex) sortAndBufferIndex;
```

Comment:
Deconstruct the tuple, so later usage is easier and the locals from L100, L101 are not needed.

Reply:
I don't mind doing it, but does it really produce better code?

Reply:
Not necessarily better code -- it should result in the same assembly code that the JIT produces.
But IMHO more readable C#-code. 

E.g. this can be avoided, as it is already done by the deconstruction.
https://github.com/dotnet/corefxlab/blob/e9f2bccdc570c2757da6e33c7ac913ecfb183062/src/Microsoft.Data/BaseColumn.cs#L100-L101

Or `nextValueAndBufferSortIndex.bufferSortIndex` looks quite long, in contrast to just `bufferSortIndex`.